homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,274 @@
+            if sensor.is_meter:
+                hass_meter_sensors.append(
+                    EnaSolarSensor(
+                        sensor, config_entry.data[CONF_NAME], enasolar.serial_no
+                    )
+                )
+            else:
+                hass_data_sensors.append(
+                    EnaSolarSensor(
+                        sensor, config_entry.data[CONF_NAME], enasolar.serial_no
+                    )
+                )
+
+    async_add_entities(hass_meter_sensors)
+    async_add_entities(hass_data_sensors)
+
+    async def async_enasolar_meters():
+        """Update the EnaSolar Meter sensors."""
+
+        if enasolar.sun_up <= datetime.now().time() <= enasolar.sun_down:
+            values = await enasolar.read_meters()
+        else:
+            values = False
+
+        for sensor in hass_meter_sensors:
+            state_unknown = False
+            if not values and (
+                (sensor.sensor.per_day_basis and date.today() > sensor.sensor.date)
+                or (
+                    not sensor.sensor.per_day_basis
+                    and not sensor.sensor.per_total_basis
+                )
+            ):
+                state_unknown = True
+            sensor.async_update_values(unknown_state=state_unknown)
+            _LOGGER.debug(
+                "Meter Sensor %s updated => %s", sensor.sensor.key, sensor.native_value
+            )
+        return values
+
+    async def async_enasolar_data():
+        """Update the EnaSolar Data sensors."""
+
+        if enasolar.sun_up <= datetime.now().time() <= enasolar.sun_down:
+            values = await enasolar.read_data()
+        else:
+            values = False
+
+        for sensor in hass_data_sensors:
+            state_unknown = False
+            if not values and (
+                (sensor.sensor.per_day_basis and date.today() > sensor.sensor.date)
+                or (
+                    not sensor.sensor.per_day_basis
+                    and not sensor.sensor.per_total_basis
+                )
+            ):
+                state_unknown = True
+            sensor.async_update_values(unknown_state=state_unknown)
+            _LOGGER.debug(
+                "Data Sensor %s updated => %s", sensor.sensor.key, sensor.native_value
+            )
+        return values
+
+    def start_update_interval(event):
+        """Start the update interval scheduling."""
+        listeners.append(
+            async_track_time_interval_backoff(
+                hass, async_enasolar_meters, SCAN_METERS_MIN_INTERVAL
+            )
+        )
+        listeners.append(
+            async_track_time_interval_backoff(
+                hass, async_enasolar_data, SCAN_DATA_MIN_INTERVAL
+            )
+        )
+
+    async def async_unload_entry(hass, config_entry):
+        """Properly cancel the scheduled update."""
+        for listener in hass.data[DOMAIN][config_entry.entry_id]["listeners"]:
+            listener()
+
```

Comment:
```suggestion
```

Reply:
So rather than storing a list of listeners, just let async_on_unload do all the work?

Reply:
Yes